rfc-drafts/stream.md

```diff
@@ -235,27 +281,156 @@ Designing such a migration feature is out of scope for this RFC.
+}
+```
+
+Examples taken from the Rust docs on [for loops and into_iter]](https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator)
+
 * `for x in iter` uses `impl IntoIterator for T`
+
+```rust
+let values = vec![1, 2, 3, 4, 5];
+
+for x in values {
+    println!("{}", x);
+}
+```
+
+Desugars to:
+
+```rust
+let values = vec![1, 2, 3, 4, 5];
+{
+    let result = match IntoIterator::into_iter(values) {
+        mut iter => loop {
+            let next;
+            match iter.next() {
+                Some(val) => next = val,
+                None => break,
+            };
+            let x = next;
+            let () = { println!("{}", x); };
+        },
+    };
+    result
+}
+```
 * `for x in &iter` uses `impl IntoIterator for &T`
 * `for x in &mut iter` uses `impl IntoIterator for &mut T`
 
+**Streams**
+
 We may want a trait similar to this for `Stream`. The `IntoStream` trait would provide a way to convert something into a `Stream`.
 
 This trait could look like this:
 
-[TO BE ADDED]
+```rust
+pub trait IntoStream where 
+    <Self::IntoStream as Stream>::Item == Self::Item,
+{
+    type Item;
+
+    type IntoStream: Stream;
+
+    fn into_stream(self) -> Self::IntoStream;
+}
+```
+
+This trait (as expressed by @taiki-e in [a comment on a draft of this RFC](https://github.com/rust-lang/wg-async-foundations/pull/15/files#r449880986)) makes it easy to write streams in combination with [async stream](https://github.com/taiki-e/futures-async-stream).
+
+```rust
+type S(usize);
+
+impl IntoStream for S {
+    type Item = usize;
+    type IntoStream: impl Stream<Item = Self::Item>;
+
+    fn into_stream(self) -> Self::IntoStream {
+        #[stream]
+        async move {
+            for i in 0..self.0 {
+                yield i;
+            }
+        }
+    }
+}   
+```
 
 ### FromStream
 
+**Iterators**
+
 Iterators have an `FromIterator` that is used to convert iterators into another type.
 
+```rust
+pub trait FromIterator<A> {
+
+    fn from_iter<T>(iter: T) -> Self
+    where
+        T: IntoIterator<Item = A>;
+}
+```
+
+It should be noted that this trait is rarely used directly, instead used through Iterator's collect method ([source](https://doc.rust-lang.org/std/iter/trait.FromIterator.html)).
+
+```rust
+pub trait Interator {
+    fn collect<B>(self) -> B
+    where
+        B: FromIterator<Self::Item>,
+    { ... }
+}
+```
+
+Examples taken from the Rust docs on [iter and collect]](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect)
+
+
+```rust
+let a = [1, 2, 3];
+
+let doubled: Vec<i32> = a.iter()
+                         .map(|&x| x * 2)
+                         .collect();
+
+```
+
+**Streams**
+
 We may want a trait similar to this for `Stream`. The `FromStream` trait would provide way to convert a `Stream` into another type.
 
 This trait could look like this:
 
-[TO BE ADDED]
+```rust
+pub trait FromStream<A> {
+    async fn from_stream<T>(stream: T) -> Self
```

Comment:
I definitely think that's possible. To clarify, are you thinking of implementing this same trait for both Streams and Iterators, or are you thinking of having separate traits with very similar functionality?

Reply:
I'm saying that `FromStream` (as defined here) and the existing `FromIterator` trait are separate, but with very similar functionality.

Thinking more about it, `FromStream` is actually more general, because an await point is allowed to suspend execution of the current function but doesn't actually have to. So many, but not all, existing impls of `FromIterator` would work for `FromStream` as well. If not for coherence with existing impls, we could do `impl<T: FromStream<Item = I>> FromIterator<I> for T`, and switch a bunch of `FromIterator` impls to be `FromStream` impls without having to repeat ourselves.

This is a somewhat similar problem to the `LendingStream`/`Stream` relationship below, but the relationship is reversed (the trait impls we have now are allowed to assume too much – that there are no suspend points – rather than too little).

This might be another possible bullet point for the lang team discussion, @nikomatsakis? (cc https://github.com/rust-lang/lang-team/issues/34)

Reply:
The way we solved this in `async-std` is by manually copying all implementations of `FromIterator` from std [(ref)](https://docs.rs/async-std/1.6.2/async_std/stream/trait.FromStream.html). When we choose to introduce `std::stream::FromStream` we could pretty much copy-paste all of the impls. I forget whether there were coherence issues (likely), but the main reason why we took this approach is because doing general conversions between sync and async code is hard to get right in practice:
- a generic conversion from `sync -> async` means operations may quietly block in the background which can lead to perf issues in runtimes
- a generic conversion from `async -> sync` means using methods such as `task::block_on` which is hard to tune, or risk deadlocks when nested (both futures-rs and async-std have had issues with this)

To help with the conversion from `Iterator` -> `Stream` we introduced [`async_std::stream::from_iter`](https://docs.rs/async-std/1.6.2/async_std/stream/fn.from_iter.html). To my knowledge all of this together has covered people's needs quite well.

Reply:
I don't think the way currently used in `async-std` is the ideal way. That way means that all third-party crates that provide the `FromIterator` implementation must be patched in order to fully support.

As far as I know, this is why `futures` implements `collect` using existing traits (`Default`, `Extend`) rather than its own. The way used in `futures` cannot currently optimize based on the size-hint of the stream, but this will be possible once https://github.com/rust-lang/rust/issues/72631 stabilizes.

If we want the collection to extend asynchronously, you can use [`Sink`](https://docs.rs/futures/0.3.5/futures/sink/trait.Sink.html) or use a loop.

> To help with the conversion from Iterator -> Stream we introduced async_std::stream::from_iter. To my knowledge all of this together has covered people's needs quite well.

This seems copy of [futures::stream::iter](https://docs.rs/futures/0.3.5/futures/stream/fn.iter.html)?

Reply:
> If we want the collection to extend asynchronously, you can use Sink or use a loop.

I don't think that's a compelling outcome; this means that patterns from sync Rust wouldn't be transferable to async Rust. Where possible I think it's incredibly valuable to aim for parity between the two. Though this kind of gets into "design philosophy", and at least in the context of this PR we don't need to find consensus.

> That way means that all third-party crates that provide the `FromIterator` implementation must be patched in order to fully support.

That's accurate, and I think that's the right outcome. In some cases `FromIterator` implementations perform blocking IO which I think is I think enough reason to not introduce a blanket conversion. For example if a function were to use `impl FromStream<Result<fs::Metadata>>` as an input, then both an iterator constructed from `std::fs::ReadDir` and a stream constructed from `async_std::fs::ReadDir` would be valid*. However the synchronous variant would likely interfere with the scheduler.

*_Admittedly it's a bit of a contrived example, but hopefully it still serves to illustrate the point._

Reply:
> and at least in the context of this PR we don't need to find consensus.

Agreed, it actually requires real async trait method to add this to std, and afaik all of the existing approaches have some problems.

> In some cases `FromIterator` implementations perform blocking IO which I think is I think enough reason to not introduce a blanket conversion.

Well, this problem also exists in `futures::stream::iter`/`async_std::stream::from_iter`.

---

Aside from it's preferable or not, even a crate that is obviously unrelated to async can't be used as a return type for `collect` without implementing it, which is necessary information when explaining/considering-adding it (I mainly think about collections).

Reply:
> Aside from it's preferable or not, even a crate that is obviously unrelated to async can't be used as a return type for `collect` without implementing it, which is necessary information when explaining/considering-adding it (I mainly think about collections).

One solution I was thinking about for this is, leave `futures`'s current `collect` method (as `collect2` or another name), even after the `FromStream` based `collect` is implemented.
(This is not an ideal solution, but it means third-party crates can provide mitigation for this, even if `std` can't solve this problem.)